### PR TITLE
Implement OAuth2 service provider functionality.

### PR DESCRIPTION
**Pull request by jwilander**
This is a rather large pull request that implements a baseline for using Mattermost as an OAuth2 service provider. In this description I will give a brief overview of our OAuth2 flow and list what is missing and needs to be added in the future.

This PR also reworks our Session-based authentication a bit to be more clear and to work better with OAuth.

**PLEASE DO NOT MERGE UNTIL MOST OF THE TEAM HAS HAD A CHANCE TO REVIEW AND HAS AGREED**

The OAuth2 spec can be found here: https://tools.ietf.org/html/rfc6749

## The OAuth2 Flow (Authorization Code flow)
Note: This is a BRIEF overview of the flow, a more detailed flow will be coming out in the future.

1. A third-party logs in to their Mattermost acount and registers an application that wishes to access user data controlled by Mattermost. This is found under "Account Settings"->"Developer"->"Applications".
2. The third-party is given a client id and client secret as the credentials for their newly registered application.
3. The third-party performs a GET request to ```https://<mattermost_domain>/oauth/authorize?response_type=code&client_id=CLIENT_ID&redirect_uri=REDIRECT_URI&state=RANDOM_STRING```
4. The user is now redirected to a Mattermost page, asking the user to allow or deny the third-party application access to their account.
5. If the user clicks 'Allow', the browser will be redirect to the ```REDIRECTURI``` provided by the third-party in step 3 with the following URL query string ```?code=AUTHORIZATION_CODE&state=RANDOM_STRING```
6. The third-party must make sure ```state``` matches the one they provided in step 3
7. The third-party will then make a POST request to exchange the ```code``` for an access token. The request will be made to ```https://<mattermost_domain>/oauth/access_token``` with the following body: 

    ```
        client_id=CLIENT_ID&
        client_secret=CLIENT_SECRET&
        code=AUTHORIZATION_CODE&
        grant_type=authorization_code&
        redirect_uri=REDIRECT_URI
    ```
8. The server will respond with the following JSON body:

    ```
    {
        "access_token": ACCESSTOKEN,
        "token_type": "bearer",
        "expires_in": 31536000, // expire time in seconds
        "scope": "", // not implemented yet
        "refresh_token": "", // not implemented yet
    }
    ```
9. The third-party can now make API calls on behalf of the user by appending ```?access_token=ACCESSTOKEN``` to API urls.

## To Do
* Have @asaadmahmoodspin clean-up the new UI elements added
* Implement the ability for app owners to edit their registered apps (will need UI spec)
* Implement refresh tokens (and add expiry to access tokens)
* Implement scope limiting
* Create detailed docs explaining our OAuth2 flows, common errors, etc.

### CODE REVIEW

File: utils/config.go [hunk 1]
```diff
@@ -21,20 +21,22 @@ const (
...
+	SiteName                   string
+	Mode                       string
+	AllowTesting               bool
+	UseSSL                     bool
+	Port                       string
+	Version                    string
+	InviteSalt                 string
+	PublicLinkSalt             string
+	ResetSalt                  string
+	TokenSalt                  string
```
Comment (coreyhulen) [inline utils/config.go]: Is TokenSalt used?

Comment (jwilander) [inline utils/config.go]: Ah, no. Not anymore. Removing..

File: web/react/pages/authorize.jsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2015 Spinpunch, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+var Authorize = require('../components/authorize.jsx');
+
+function setupAuthorizePage(teamName, appName, responseType, clientId, redirectUri, scope, state) {
```
Comment (coreyhulen) [inline web/react/pages/authorize.jsx]: This should use the new mechanism of {{.Props}}

**Comment by coreyhulen** [comment]
Was "oauth" added as a team reversed word?

**Comment by coreyhulen** [comment]
minor nits can be addressed afterwards.... +1

**Comment by jwilander** [comment]
Added Jira tickets to add "oauth" as reserved word and use {{.Props}} in authorize.jsx

**Comment by crspeller** [comment]
+1 LETS DO THIS

**Pull request merged**